@xen-orchestra/proxy/src/app/_debounceResource.js

```diff
@@ -1,35 +1,54 @@
 import Disposable from 'promise-toolbox/Disposable'
+import { createLogger } from '@xen-orchestra/log/dist'
 
-export async function debounceResource(pDisposable, hooks, delay = 0) {
-  if (delay === 0) {
-    return pDisposable
-  }
+import { asyncMap } from '../_asyncMap'
+
+const { warn } = createLogger('xo:proxy:debounceResource')
+
+export const createDebounceResource = defaultDelay => {
+  const flushers = new Set()
+  async function debounceResource(pDisposable, delay = defaultDelay) {
+    if (delay === 0) {
+      return pDisposable
+    }
 
-  const disposable = await pDisposable
+    const disposable = await pDisposable
 
-  let timeoutId
-  const disposeWrapper = () => {
-    if (timeoutId !== undefined) {
-      clearTimeout(timeoutId)
-      timeoutId = undefined
-      hooks.removeListener('stop', onStop)
+    let timeoutId
+    const disposeWrapper = async () => {
+      if (timeoutId !== undefined) {
+        clearTimeout(timeoutId)
+        timeoutId = undefined
+        flushers.delete(flusher)
 
-      return disposable.dispose()
+        try {
+          await disposable.dispose()
+        } catch (error) {
+          warn(error)
+        }
+      }
     }
-  }
 
-  const onStop = () => {
-    const shouldDisposeNow = timeoutId !== undefined
-    if (shouldDisposeNow) {
-      return disposeWrapper()
-    } else {
-      // will dispose ASAP
-      delay = 0
+    const flusher = () => {
+      const shouldDisposeNow = timeoutId !== undefined
+      if (shouldDisposeNow) {
+        return disposeWrapper()
+      } else {
+        // will dispose ASAP
+        delay = 0
+      }
     }
+    flushers.add(flusher)
+
+    return new Disposable(disposable.value, () => {
+      timeoutId = setTimeout(disposeWrapper, delay)
+    })
+  }
+  debounceResource.flushAll = () => {
+    const currentFlushers = [...flushers]
```

Comment:
You don't need to do this with `asyncMap` :slightly_smiling_face:

Reply:
I need to create a copy of the `Set` because;

```js
const foo = new Set([1, 2])
const bar = foo
foo.clear()
console.log(bar) => Set(0){}
```

Reply:
You can avoid this by iterating before clearing:

```js
const promise = asyncMap(set)
set.clear()
return promise
```